espressif-devcon/2023/2023.md

```diff
@@ -17,15 +17,180 @@ headingDivider: 1
+            // When data becomes available, `wake` will be called, and the
+            // user of this `Future` will know to call `poll` again and
+            // receive data.
+            self.socket.set_readable_callback(wake);
+            Poll::Pending
+        }
+    }
+}
+```
+
+# Why use async?
+
+- Single-threaded concurrency (multitasking), doing multiple things at once without needing threads
+- Can run on a single stack, great for microcontrollers
+- Composability of async/await instead of manually writing state machines.
+
+## Busy Loop
+
+- repeatedly checks for condition to be true before proceeding
+- simple to implement, very inefficient
+
+```rust
+    let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
+    let mut led = io.pins.gpio7.into_push_pull_output();
+    let button = io.pins.gpio9.into_pull_up_input();
+
+    loop {
+        if button.is_high().unwrap() {
+            led.set_high().unwrap();
+        } else {
+            led.set_low().unwrap();
+        }
+    }
+```
+
+## Interrupt
+
+- hardware signal that interrupts the normal flow of program's execution
+- more code required, harder to write and read the code, efficient
+
+```rust
+    static BUTTON: Mutex<RefCell<Option<Gpio9<Input<PullDown>>>>> = Mutex::new(RefCell::new(None));
+    // inside main()
+    let mut led = io.pins.gpio7.into_push_pull_output();
+
+    // Set GPIO9 as an input
+    let mut button = io.pins.gpio9.into_pull_down_input();
+    button.listen(Event::FallingEdge);
+
+    critical_section::with(|cs| BUTTON.borrow_ref_mut(cs).replace(button));
+
+    interrupt::enable(peripherals::Interrupt::GPIO, interrupt::Priority::Priority3).unwrap();
+
+    unsafe {
+        riscv::interrupt::enable();
+    }
+
+    let mut delay = Delay::new(&clocks);
+    loop {
+        led.toggle().unwrap();
+        delay.delay_ms(500u32);
+    }
+    // end of main()
+
+    #[interrupt]
+    fn GPIO() {
+        critical_section::with(|cs| {
+            println!("GPIO interrupt");
+            BUTTON
+                .borrow_ref_mut(cs)
+                .as_mut()
+                .unwrap()
+                .clear_interrupt();
+        });
+    }
+```
+
+## Async
+
+- higher-level approach used to handle multiple tasks without resorting the busy wait
+- syntax-wise it's similar to the `busy loop` with `async/.await` but it is more efficient
+
+```rust
+    #[embassy_executor::task]
+    async fn ping(mut pin: Gpio9<Input<PullDown>>) {
+        loop {
+            pin.wait_for_rising_edge().await.unwrap();
+            Timer::after(Duration::from_millis(100)).await;
+        }
+    }
+
+    static EXECUTOR: StaticCell<Executor> = StaticCell::new();
+
+    // inside main()
+    let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
+    // GPIO 9 as input
+    let input = io.pins.gpio9.into_pull_down_input();
+
+    // Async requires the GPIO interrupt to wake futures
+    esp32c3_hal::interrupt::enable(
```

Comment:
Let's remove this interrupt enable block, I don't think it adds much to what we're trying to convey with the example.